src/wait.rs

```diff
@@ -0,0 +1,155 @@
+use crate::{elements::Element, error, Client, Locator};
+use core::fmt;
+use futures_util::TryFutureExt;
+use std::{
+    error::Error,
+    future::Future,
+    pin::Pin,
+    time::{Duration, Instant},
+};
+
+const DEFAULT_TIMEOUT: Duration = Duration::from_secs(30);
+const DEFAULT_PERIOD: Duration = Duration::from_millis(250);
+
+/// Used for setting up a wait operation on the client.
+#[derive(Debug)]
+pub struct Wait<'c> {
+    client: &'c mut Client,
+    timeout: Option<Duration>,
+    period: Duration,
+}
+
+impl<'c> Wait<'c> {
+    /// Create a new wait operation from a client.
+    pub fn new(client: &'c mut Client) -> Self {
+        Self {
+            client,
+            timeout: Some(DEFAULT_TIMEOUT),
+            period: DEFAULT_PERIOD,
+        }
+    }
+
+    /// Set the timeout until the operation should wait.
+    pub fn until(mut self, timeout: Duration) -> Self {
+        self.timeout = Some(timeout);
+        self
+    }
+
+    /// Wait forever.
+    pub fn forever(mut self) -> Self {
+        self.timeout = None;
+        self
+    }
+
+    /// Sets the period to delay checks.
+    pub fn every(mut self, period: Duration) -> Self {
+        self.period = period;
+        self
+    }
+
+    /// Wait until a condition exists or a timeout is hit.
+    pub async fn on<T, F>(self, mut f: F) -> Result<T, WaitError>
+    where
+        F: for<'f> FnMut(
+            &'f mut Client,
+        )
+            -> Pin<Box<dyn Future<Output = Result<Option<T>, error::CmdError>> + 'f>>,
```

Comment:
I don't think you need `Pin<Box<dyn` here. Does this work?
```suggestion
    pub async fn on<T, F, Fut>(self, mut f: F) -> Result<T, WaitError>
    where
        F: for<'f> FnMut(&'f mut Client) -> Fut + 'f,
        Fut: Future<Output = Result<Option<T>, error::CmdError>
```

Reply:
In fact, you may even be able to remove all mention of `'f`

Reply:
Yes it works :grinning: 

Dropping either the `'f` lifetime, or using a extra type argument for the future (so that I don't need to use a PinBox) both makes me run into lifetime issue, as the compiler can't figure it out on its own.

The goal is to use the client inside the "on" function. So the function must be valid for the lifetime of the client only, but not necessarily for `'static'`, what the compiler falls back to (complaining that the function is not valid for `'static`).

Using a [Higher-Rank Trait Bound](https://doc.rust-lang.org/nomicon/hrtb.html#higher-rank-trait-bounds-hrtbs) seems to do the trick.

Having said that, maybe there is a better way do deal with this. I would be happy to learn about it :grin:.

Reply:
Interesting, that's unfortunate. Well, let's keep in `PinBox` then :sweat_smile: